### PR TITLE
feat: coupons recurring part

### PR DESCRIPTION
**Pull request by lovrocolic**
## Context

This change includes frequency logic into coupons.

## Description

We introduce `frequency` and `frequency_duration` fields into coupons.

Applied coupon termination should happen (for recurring coupons) when frequency duration becomes zero, for both fixed amount and percentage coupons.
Also, on recurring coupons we don't have to track remaining amount on coupon anymore. Each time we should just try to use as much as possible from the coupon.

### CODE REVIEW

File: app/services/credits/applied_coupon_service.rb [hunk 1]
```diff
@@ -40,9 +41,13 @@ def already_applied?
...
     def compute_amount
-      return invoice.amount_cents if remaining_amount > invoice.amount_cents
+      if applied_coupon.coupon.percentage?
+        discounted_value = invoice.amount_cents * applied_coupon.percentage_rate.fdiv(100)
 
-      remaining_amount
+        return (discounted_value >= invoice.amount_cents) ? invoice.amount_cents : discounted_value.round
+      end
+
+       applied_coupon.recurring? ? calculate_recurring_amount : calculate_one_time_coupon_amount
```
Comment (jdenquin) [inline app/services/credits/applied_coupon_service.rb:50]: Layout/IndentationConsistency: Inconsistent indentation detected. (https://rubystyle.guide#spaces-indentation, https://edgeguides.rubyonrails.org/contributing_to_ruby_on_rails.html#follow-the-coding-conventions)

File: db/migrate/20220922105251_add_frequency_and_frequency_duration_to_coupons.rb [hunk 2]
```diff
@@ -0,0 +1,13 @@
+class AddFrequencyAndFrequencyDurationToCoupons < ActiveRecord::Migration[7.0]
```
Comment (jdenquin) [inline db/migrate/20220922105251_add_frequency_and_frequency_duration_to_coupons.rb:1]: Style/FrozenStringLiteralComment: Missing frozen string literal comment.

File: spec/services/credits/applied_coupon_service_spec.rb [hunk 3]
```diff
@@ -97,5 +97,133 @@
...
+      end
+
+      it 'does not terminate the applied coupon' do
+        result = credit_service.create
+
+        expect(result).to be_success
+        expect(applied_coupon.reload).not_to be_terminated
+      end
+
+      context 'and coupon amount is higher than invoice amount' do
```
Comment (jdenquin) [inline spec/services/credits/applied_coupon_service_spec.rb:153]: RSpec/ContextWording: Context description should match /^when\b/, /^with\b/, or /^without\b/. (https://rspec.rubystyle.guide/#context-descriptions, https://www.rubydoc.info/gems/rubocop-rspec/RuboCop/Cop/RSpec/ContextWording)

File: spec/services/credits/applied_coupon_service_spec.rb [hunk 4]
```diff
@@ -97,5 +97,133 @@
...
+    context 'when coupon is recurring and percentage' do
+      let(:coupon) { create(:coupon, frequency: 'recurring', frequency_duration: 3, coupon_type: 'percentage') }
+
+      let(:applied_coupon) do
+        create(
+          :applied_coupon,
+          coupon: coupon,
+          frequency: 'recurring',
+          frequency_duration: 3,
+          percentage_rate: 20.00
```
Comment (jdenquin) [inline spec/services/credits/applied_coupon_service_spec.rb:174]: Style/TrailingCommaInArguments: Put a comma after the last parameter of a multiline method call. (https://rubystyle.guide#no-trailing-params-comma)

File: spec/services/credits/applied_coupon_service_spec.rb [hunk 5]
```diff
@@ -97,5 +97,133 @@
...
+      end
+
+      it 'does not terminate the applied coupon' do
+        result = credit_service.create
+
+        expect(result).to be_success
+        expect(applied_coupon.reload).not_to be_terminated
+      end
+
+      context 'and frequency duration becomes zero' do
```
Comment (jdenquin) [inline spec/services/credits/applied_coupon_service_spec.rb:197]: RSpec/ContextWording: Context description should match /^when\b/, /^with\b/, or /^without\b/. (https://rspec.rubystyle.guide/#context-descriptions, https://www.rubydoc.info/gems/rubocop-rspec/RuboCop/Cop/RSpec/ContextWording)

File: spec/services/credits/applied_coupon_service_spec.rb [hunk 6]
```diff
@@ -97,5 +97,133 @@
...
+        let(:applied_coupon) do
+          create(
+            :applied_coupon,
+            coupon: coupon,
+            frequency: 'recurring',
+            frequency_duration: 1,
+            percentage_rate: 20.00
```
Comment (jdenquin) [inline spec/services/credits/applied_coupon_service_spec.rb:204]: Style/TrailingCommaInArguments: Put a comma after the last parameter of a multiline method call. (https://rubystyle.guide#no-trailing-params-comma)

**Comment by lovrocolic** [comment]
Will open new PR for theses changes

**Pull request closed without merging**